packages/firestore/src/api/user_data_converter.ts

```diff
@@ -536,7 +536,9 @@ export class UserDataConverter {
       if (input instanceof Array) {
         // TODO(b/34871131): Include the path containing the array in the error
         // message.
-        if (context.arrayElement) {
+        // Nested arrays are allowed when making IN queries, so we make the
+        // exception here.
+        if (context.arrayElement && context.methodName !== 'Query.where') {
```

Comment:
This works in JS but is a little fragile (relies on string equality of magic vales) and won't be portable to other platforms since they don't have access to the methodName. [The only reason we have the methodName in JS is because we try to include it in the error message since (historically at least) in some cases you may not have a meaningful callstack to show you what method the error came from.]

I think it'd probably be better to use `context.dataSource` which is a more formalized way to know where the data came from and how we should parse it.  We could probably get away with `context.dataSource !== UserDataSource.Argument` here.  This means we will disable the nested-array validation in more cases, but I think it's probably okay.  If we want to keep it more targeted, we could introduce `UserDataSource.ArrayArgument`, indicating that we're parsing an array and so "nested" arrays is okay.

The other option I had mentioned previously would be to just parse the elements of the array individually for 'in' filters, e.g. [here](https://github.com/firebase/firebase-js-sdk/blob/efc0a78803990eb01888b7b2121c15377c4007b1/packages/firestore/src/api/database.ts#L1488) we would do:
```
        fieldValue = new ArrayValue((value as unknown[]).map(val => this.firestore._dataConverter.parseQueryValue('Query.where', value)));
```

I don't have strong feelings, so whichever of these sounds good to you works for me.